### PR TITLE
[ROBO-3381] split and merge poc 

### PR DESCRIPTION
**Pull request by mihainradu**

### CODE REVIEW

File: src/UiPath.Workflow.Runtime/Statements/Flowchart.cs [hunk 1]
```diff
@@ -179,199 +43,166 @@ protected override void CacheMetadata(NativeActivityMetadata metadata)
...
+                AddValidationError($"Node has multiple splits incoming branches. Please precede with a Merge node.", node, splits);
         }
+        void ValidateSplit(FlowSplit split)
+        {
+            HashSet<FlowMerge> allMerges = new();
+            foreach (var branch in split.Branches)
+            {
+                var merges = GetMerges(branch).Distinct().ToList();
+                allMerges.AddRange(merges);
+            }
```
Comment (danutboanta) [inline src/UiPath.Workflow.Runtime/Statements/Flowchart.cs]: You can simplify by getting all the merges for a split, e.g.:
`var allMerges = GetMerges(split)`

and GetMerges can be:

`
List<FlowMerge> GetMerges(FlowSplit flowSplit)
{
    return (
        from nodeInfo in _staticBranchesByNode
        where nodeInfo.Key is FlowMerge
        where nodeInfo.Value.GetTop().Contains(flowSplit)
        select nodeInfo.Key as FlowMerge
        ).ToList();
}
`

File: src/UiPath.Workflow.Runtime/Statements/Flowchart.cs [hunk 2]
```diff
@@ -179,199 +43,166 @@ protected override void CacheMetadata(NativeActivityMetadata metadata)
...
+                AddValidationError("A merge must be preceded by a Split node.", node);
+        }
+        void AddValidationError(string message, FlowNode node, IEnumerable<FlowNode> otherNodes = null)
+        {
+            metadata.AddValidationError(new ValidationError(message)
+            {
+                SourceDetail = new[] { node }.Concat(otherNodes ?? Array.Empty<FlowNode>()).ToArray()
+            });
+        }
+        List<FlowMerge> GetMerges(FlowNode flowNode)
```
Comment (danutboanta) [inline src/UiPath.Workflow.Runtime/Statements/Flowchart.cs]: Change to return the merges for a split

File: src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                    ).ToList();
+            AddStack(result);
+        }
+
+        public void PropagateStack(StaticNodeStackInfo preStacks)
+        {
+            AddStack(preStacks.SplitsStack);
+        }
+
+        private void AddStack(List<List<FlowSplit>> stack)
```
Comment (danutboanta) [inline src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs]: Name proposal: AddUniqueStack

File: src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        public void AddPop(StaticNodeStackInfo popFrom)
+        {
+            var result = popFrom.SplitsStack.Select(s => Enumerable.Reverse(s).Skip(1).Reverse().ToList()).ToList();
+            AddStack(result);
+        }
+
+        private bool HasStack(List<FlowSplit> stack)
+            => SplitsStack.Any(b => stack.Count == b.Count && stack.All(p => b.Contains(p)));
```
Comment (danutboanta) [inline src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs]: => SplitsStack.Any(s => s.SequenceEqual(stack));

File: src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+        internal bool IsOnStack(StaticNodeStackInfo toConfirm)
```
Comment (danutboanta) [inline src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs]: Not needed, if you rewrite the GetMerges method

File: src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            {
+                if (HasStack(pre))
+                    continue;
+                SplitsStack.Add(pre);
+            }
+        }
+
+        public void AddPop(StaticNodeStackInfo popFrom)
+        {
+            var result = popFrom.SplitsStack.Select(s => Enumerable.Reverse(s).Skip(1).Reverse().ToList()).ToList();
```
Comment (danutboanta) [inline src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs]: Quite difficult to understand

File: src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        private List<List<FlowSplit>> SplitsStack { get; init; } = new();
+        public static StaticNodeStackInfo EmptyStack { get; } = new StaticNodeStackInfo(){SplitsStack = new() { new() } };
+
+        public HashSet<FlowSplit> GetTop()
+        {
+            return new (SplitsStack.Select(b => b.LastOrDefault()));
+        }
+
+        public void Push(FlowSplit newSplit, StaticNodeStackInfo splitsStackInfo)
```
Comment (danutboanta) [inline src/UiPath.Workflow.Runtime/Statements/Flowchart.StaticNodeStackInfo.cs]: Naming: Push vs AddPop

File: src/UiPath.Workflow.Runtime/Statements/Flowchart.cs [hunk 8]
```diff
@@ -179,199 +43,166 @@ protected override void CacheMetadata(NativeActivityMetadata metadata)
...
-            if (context.IsCancellationRequested)
+            Stack<FlowNode> toVisit = new();
+            toVisit.Push(StartNode);
+            GetStaticSplitsStack(StartNode).PropagateStack(StaticNodeStackInfo.EmptyStack);
+            while (toVisit.TryPop(out var current))
             {
-                Fx.Assert(completedInstance != null, "cannot request cancel if we never scheduled any children");
-                // we are done but the last child didn't complete successfully
-                if (completedInstance.State != ActivityInstanceState.Closed)
+                if (WasVisited(current))
```
Comment (danutboanta) [inline src/UiPath.Workflow.Runtime/Statements/Flowchart.cs]: WasVisited is based on node.Flowchart != null. 
I did not understand if the new nodes are created every time before CacheMetadata

Comment (mihainradu) [reply src/UiPath.Workflow.Runtime/Statements/Flowchart.cs]: ?

**Pull request closed without merging**